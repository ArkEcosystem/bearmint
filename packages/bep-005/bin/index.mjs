#!/usr/bin/env node
import { join } from 'node:path'

import { globby } from 'globby'
import meow from 'meow'

process.env.NODE_ENV = 'test'

const { flags } = meow({
	flags: {
		config: {
			type: 'string',
		},
	},
	importMeta: import.meta,
})

const { files } = (await import(join(process.cwd(), flags.config))).default
const tests = (await globby(files)).map((path) => join(process.cwd(), path))

for (const test of tests) {
	await import(test)
}
