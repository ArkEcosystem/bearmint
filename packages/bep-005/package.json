{
	"name": "@bearmint/bep-005",
	"version": "0.0.0",
	"private": true,
	"description": "This library is the official reference implementation for BEP-005 (https://bep.dev/005)",
	"keywords": [
		"bearmint",
		"tendermint",
		"blockchain",
		"framework",
		"modular"
	],
	"homepage": "https://github.com/bearmint/bearmint",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/bearmint/bearmint.git"
	},
	"funding": "https://funding.bearmint.com/",
	"license": "MIT",
	"author": "Bearmint Mining Company <hello@bearmint.com> (https://bearmint.com/)",
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"bin": {
		"roadrunner": "./bin/index.mjs"
	},
	"files": [
		"bin",
		"distribution"
	],
	"scripts": {
		"audit": "pnpm audit --fix --prod",
		"build": "pnpm run clean && tsc",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rm -rf distribution",
		"depcheck": "depcheck --ignores=@bearmint/bep-004,@bearmint/bep-005,@bearmint/bep-031,@bearmint/bep-039,globby,meow",
		"deps:major": "ncu -u --target=latest",
		"deps:minor": "ncu -u --target=minor",
		"deps:patch": "ncu -u --target=patch"
	},
	"dependencies": {
		"@bearmint/bep-013": "workspace:^",
		"@bearmint/bep-018": "workspace:^",
		"@bearmint/bep-116": "workspace:^",
		"@sinonjs/fake-timers": "9.1.2",
		"ansi-styles": "6.2.1",
		"concordance": "5.0.4",
		"globby": "13.1.2",
		"lodash.get": "4.4.2",
		"lodash.has": "4.5.2",
		"meow": "^11.0.0",
		"nock": "13.2.9",
		"pretty-ms": "8.0.0",
		"sinon": "14.0.1",
		"strip-ansi": "7.0.1",
		"vitest": "^0.24.5"
	},
	"devDependencies": {
		"@types/node": "18.11.9",
		"@types/sinon": "10.0.13",
		"@types/sinonjs__fake-timers": "8.1.2"
	},
	"engines": {
		"node": ">=18",
		"pnpm": ">=7.9.0"
	},
	"publishConfig": {
		"access": "restricted"
	}
}
