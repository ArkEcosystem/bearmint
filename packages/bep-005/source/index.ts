export * from './expect.js'
export * from './fake.js'
export * from './nock.js'
export * from './spy.js'
export * from './stub.js'
