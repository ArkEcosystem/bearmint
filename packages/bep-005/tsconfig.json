{
	"compilerOptions": {
		"typeRoots": ["node_modules/@types"],
		"outDir": "distribution"
	},
	"exclude": ["node_modules", "test/**/**.ts"],
	"extends": "../../config/tsconfig.json",
	"include": ["source/**/**.ts"]
}
