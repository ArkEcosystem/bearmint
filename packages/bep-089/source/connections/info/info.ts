import { semver } from '@bearmint/bep-009'
import type { ABCIController, Cradle } from '@bearmint/bep-013'
import type { abci } from '@bearmint/bep-018'

/**
 * @see https://github.com/tendermint/tendermint/blob/main/spec/abci/abci%2B%2B_methods.md#info
 */
export function makeInfo(cradle: Cradle): ABCIController<abci.RequestInfo, abci.ResponseInfo> {
	return {
		async execute(_request) {
			const base: Partial<abci.ResponseInfo> = {
				// The application protocol version
				appVersion: BigInt(semver.toNumber(cradle.ApplicationManifest.protocolVersion)),
				// The application semantic version
				version: cradle.ApplicationManifest.semanticVersion,
			}

			try {
				return {
					...base,
					data: cradle.ApplicationManifest.name,
					lastBlockAppHash: await cradle.CommittedState.getCommittedBlockAppHash(),
					lastBlockHeight: await cradle.CommittedState.getCommittedBlockNumber(),
				}
			} catch {
				return {
					...base,
					lastBlockAppHash: new Uint8Array(0),
					lastBlockHeight: BigInt(0),
				}
			}
		},
	}
}
