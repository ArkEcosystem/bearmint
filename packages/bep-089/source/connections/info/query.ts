import type { ABCIController, Cradle } from '@bearmint/bep-013'
import type { abci } from '@bearmint/bep-018'
import { ErrorCode } from '@bearmint/bep-109'
import { serializeError } from 'serialize-error'

export function makeQuery(cradle: Cradle): ABCIController<abci.RequestQuery, abci.ResponseQuery> {
	return {
		async execute(request) {
			try {
				const path: string[] = request.path.slice(1).split('/')

				for (const queryHandlerFactory of cradle.QueryHandlerRegistry.values()) {
					const queryHandler = cradle.Container.build(queryHandlerFactory)

					if (await queryHandler.isMatch({ data: request.data, path })) {
						return await queryHandler.execute({ data: request.data, path })
					}
				}

				return {
					code: 1,
					info: ErrorCode.TM_QUERY_UNKNOWN,
				}
			} catch (error) {
				return {
					code: 1,
					info: `${ErrorCode.TM_QUERY_FAULTY}: ${JSON.stringify(serializeError(error))}`,
				}
			}
		},
	}
}
