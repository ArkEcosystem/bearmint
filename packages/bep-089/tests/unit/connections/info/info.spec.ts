import { expect, describe, it } from 'vitest'
import { extendExpect } from '@bearmint/bep-005'
extendExpect(expect)

import { makeInfo } from '../../../../source/connections/info/info.js'

describe('makeInfo', () => {
	it('it should respond without any errors', async () => {
		const commitedAppHash = Buffer.from('deadbeef')
		const commitedBlockNumber = BigInt(1)

		expect(
			await makeInfo({
				ApplicationManifest: {
					semanticVersion: '1.0.0',
					name: 'Bearmint',
					protocolVersion: '1.0.0',
				},
				CommittedState: {
					getAppHash() {
						return commitedAppHash
					},
					getCommittedBlockNumber() {
						return commitedBlockNumber
					},
					getMultiStore() {
						return {
							roots() {
								return {
									world: Buffer.alloc(8),
								}
							},
						}
					},
				},
			}).execute({}),
		).toMatchSnapshot()
	})
})
