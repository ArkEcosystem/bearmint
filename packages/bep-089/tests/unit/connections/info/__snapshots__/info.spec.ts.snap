// Vitest Snapshot v1

exports[`makeInfo > it should respond without any errors 1`] = `
{
  "appVersion": 10000n,
  "data": "Bearmint",
  "lastBlockAppHash": {
    "data": [
      100,
      101,
      97,
      100,
      98,
      101,
      101,
      102,
    ],
    "type": "Buffer",
  },
  "lastBlockHeight": 1n,
  "version": "1.0.0",
}
`;
